### ISSUE TITLE
CommitAbilityCooldown not updating to GSCUserWidget OnCooldown Start

### ISSUE THREAD

**Issue by MojoAA13** [GASCompanion/GASCompanion-Documentation#31]
**Describe the bug**
Calling CommitAbilityCooldown commits the cooldown, but does not cause GSCUser widget OnCooldownStart to trigger

**To Reproduce**
What did you do?

It is best to reproduce in a blank project (with GAS Companion installed in engine, not project). If possible provide a link to this project either as a github repo or google drive link to a zip file (or similar platforms) and a list of clear reproduction steps.

**Expected behavior**

**Actual behavior**
What happened instead?

**Screenshots**

**Environment (please complete the following information):**
- Engine version: 4.26 / 4.27 / 5.0 early access
- GAS Companion Plugin version: 3.2.0 (please be specific, and try latest release. You can find this information in the Plugins window within the Editor)
- OS: Windows / Mac / Linux

**Additional context**

**Comment by mklabs** [comment]
Thank you for the report.

Indeed, this was briefly discussed on discord as well.

https://discord.com/channels/198737843743358985/1266500914995396620

Now thinking about it, your report here is slightly different than the feedback you gave on discord before. Your issue was related to duration and time remaining parameters for the GSCUserWidget::OnCooldownStart  event to be at 0 in that scenario.

If you are not seeing any call at all for OnCooldownStart, one thing to make sure is to check the boolean flag `BroadcastCommitEvent` is set to true, it's false by default if you don't manually check it when using CommitAbilityCooldown() node.

---

For duration and time remaining values not calculated properly, I believe this is due to engine broadcasting the event before commit ability cooldown.

https://github.com/EpicGames/UnrealEngine/blob/5.5.1-release/Engine/Plugins/Runtime/GameplayAbilities/Source/GameplayAbilities/Private/Abilities/GameplayAbility.cpp#L1383-L1394

Unlike the CommitAbility node where the delegate is invoked by the engine right after commit happens. With CommitAbilityCooldown or CommitAbilityCost, the call to the delegate happens just before, that's why the event you have in GSCUserWidget happens prior to that (and is likely the reason for duration and time remaining values to be at 0).

For context, here is the the fix I tested to workaround this issue. This might eventually be implemented, I don't really like it but only thing I could do on GAS Companion side.

Basically waiting for a frame before invoking the event, so that the cooldown is committed before trying to read duration and time remaining values.

```cpp
    // AbilitySystemComponent->AbilityCommittedCallbacks.AddUObject(this, &UGSCUserWidget::OnAbilityCommitted);
    AbilitySystemComponent->AbilityCommittedCallbacks.AddWeakLambda(this, [WeakThis = MakeWeakObjectPtr(this)](UGameplayAbility* ActivatedAbility)
    {
        if (!WeakThis.IsValid() || !WeakThis->GetWorld())
        {
            return;
        }

        WeakThis->GetWorld()->GetTimerManager().SetTimerForNextTick([WeakThis, ActivatedAbility]()
        {
            WeakThis->OnAbilityCommitted(ActivatedAbility);
        });
    });
```

Documenting it here in case I need to refer back to it when landing a fix.
